File: BrickKit/bricks/src/androidTest/java/com/wayfair/brickkit/BrickSpanSizeLookupTest.java
/**
 * Copyright © 2017 Wayfair. All rights reserved.
 */
package com.wayfair.brickkit;

import android.content.Context;
import android.support.test.InstrumentationRegistry;
import android.support.test.runner.AndroidJUnit4;

import com.wayfair.brickkit.brick.BaseBrick;
import com.wayfair.brickkit.size.BrickSize;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.LinkedList;

import static org.junit.Assert.assertEquals;
import static org.mockito.Matchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@RunWith(AndroidJUnit4.class)
public class BrickSpanSizeLookupTest {
    private static final int DEFAULT_SPANS = 1;
    private static final int SPANS = 5;
    private static final int POSITION = 4;

    private BrickDataManager manager;
    private LinkedList<BaseBrick> bricks;
    private BrickSpanSizeLookup brickSpanSizeLookup;

    @Before
    public void setup() {
        manager = mock(BrickDataManager.class);

        bricks = mock(LinkedList.class);

        when(manager.getRecyclerViewItems()).thenReturn(bricks);

        brickSpanSizeLookup = new BrickSpanSizeLookup(InstrumentationRegistry.getTargetContext(), manager);
    }

    @Test
    public void testDefaultSpanSize() {
        when(bricks.get(POSITION)).thenReturn(null);

        assertEquals(DEFAULT_SPANS, brickSpanSizeLookup.getSpanSize(POSITION));
    }

    @Test
    public void testBrickSpanSize() {
        BrickSize brickSize = mock(BrickSize.class);
        when(brickSize.getSpans(any(Context.class))).thenReturn(SPANS);

        BaseBrick brick = mock(BaseBrick.class);
        when(brick.getSpanSize()).thenReturn(brickSize);

        when(bricks.get(POSITION)).thenReturn(brick);

        when(manager.getRecyclerViewItems()).thenReturn(bricks);
        assertEquals(SPANS, brickSpanSizeLookup.getSpanSize(POSITION));
    }
}

File: BrickKit/bricks/src/androidTest/java/com/wayfair/brickkit/BrickSpanSizeLookupTest.kt
/**
 * Copyright © 2017-2021 Wayfair. All rights reserved.
 */
package com.wayfair.brickkit

import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import org.mockito.kotlin.any
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever
import com.wayfair.brickkit.brick.BaseBrick
import com.wayfair.brickkit.size.BrickSize
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class BrickSpanSizeLookupTest {

    private var manager: BrickDataManager = mock()
    private lateinit var bricks: MutableList<BaseBrick>
    private lateinit var brickSpanSizeLookup: BrickSpanSizeLookup

    @Before
    fun setup() {
        bricks = mutableListOf()
        whenever(manager.recyclerViewItems).thenReturn(bricks)

        brickSpanSizeLookup = BrickSpanSizeLookup(ApplicationProvider.getApplicationContext(), manager)
    }

    @Test
    fun testDefaultSpanSize_nullItem() {
        assertEquals(DEFAULT_SPANS, brickSpanSizeLookup.getSpanSize(0))
    }

    @Test
    fun testDefaultSpanSize_nullSpanSize() {
        val brick = mock<BaseBrick>()
        whenever(brick.spanSize).thenReturn(null)

        bricks.add(brick)

        assertEquals(DEFAULT_SPANS, brickSpanSizeLookup.getSpanSize(0))
    }

    @Test
    fun testBrickSpanSize() {
        val brickSize = mock<BrickSize>()
        whenever(brickSize.getSpans(any())).thenReturn(SPANS)
        val brick = mock<BaseBrick>()
        whenever(brick.spanSize).thenReturn(brickSize)

        bricks.add(brick)

        assertEquals(SPANS, brickSpanSizeLookup.getSpanSize(0))
    }

    companion object {
        private const val DEFAULT_SPANS = 1
        private const val SPANS = 5
    }
}
